src/Temporalio/Bridge/src/metric.rs

```diff
@@ -186,3 +196,295 @@ impl From<&MetricIntegerOptions> for metrics::MetricParameters {
+
+type CustomMetricMeterMeterFreeCallback = unsafe extern "C" fn(meter: *const CustomMetricMeter);
+
+/// No parameters in the callbacks below should be assumed to live beyond the
+/// callbacks unless they are pointers to things that were created lang-side
+/// originally. There are no guarantees on which thread these calls may be
+/// invoked on.
+#[repr(C)]
+pub struct CustomMetricMeter {
+    pub metric_integer_new: CustomMetricMeterMetricIntegerNewCallback,
+    pub metric_integer_free: CustomMetricMeterMetricIntegerFreeCallback,
+    pub metric_integer_update: CustomMetricMeterMetricIntegerUpdateCallback,
+    pub attributes_new: CustomMetricMeterAttributesNewCallback,
+    pub attributes_free: CustomMetricMeterAttributesFreeCallback,
+    pub meter_free: CustomMetricMeterMeterFreeCallback,
+}
+
+#[repr(C)]
+pub struct CustomMetricAttribute {
+    pub key: ByteArrayRef,
+    pub value: CustomMetricAttributeValue,
+    pub value_type: MetricAttributeValueType,
+}
+
+#[repr(C)]
+pub union CustomMetricAttributeValue {
+    string_value: CustomMetricAttributeValueString,
+    int_value: i64,
+    float_value: f64,
+    bool_value: bool,
+}
+
+// We create this type because we want it to implement Copy
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct CustomMetricAttributeValueString {
+    data: *const u8,
+    size: libc::size_t,
+}
+
+#[derive(Debug)]
+pub struct CustomMetricMeterRef {
+    meter_impl: Arc<CustomMetricMeterImpl>,
+}
+
+unsafe impl Send for CustomMetricMeterRef {}
+unsafe impl Sync for CustomMetricMeterRef {}
+
+impl metrics::CoreMeter for CustomMetricMeterRef {
+    fn new_attributes(&self, attribs: metrics::NewAttributes) -> metrics::MetricAttributes {
+        self.build_attributes(None, attribs)
+    }
+
+    fn extend_attributes(
+        &self,
+        existing: metrics::MetricAttributes,
+        attribs: metrics::NewAttributes,
+    ) -> metrics::MetricAttributes {
+        self.build_attributes(Some(existing), attribs)
+    }
+
+    fn counter(&self, params: metrics::MetricParameters) -> Arc<dyn metrics::Counter> {
+        Arc::new(self.new_metric_integer(params, MetricIntegerKind::Counter))
+    }
+
+    fn histogram(&self, params: metrics::MetricParameters) -> Arc<dyn metrics::Histogram> {
+        Arc::new(self.new_metric_integer(params, MetricIntegerKind::Histogram))
+    }
+
+    fn gauge(&self, params: metrics::MetricParameters) -> Arc<dyn metrics::Gauge> {
+        Arc::new(self.new_metric_integer(params, MetricIntegerKind::Gauge))
+    }
+}
+
+impl CustomMetricMeterRef {
+    pub fn new(meter: *const CustomMetricMeter) -> CustomMetricMeterRef {
+        CustomMetricMeterRef {
+            meter_impl: Arc::new(CustomMetricMeterImpl(meter)),
+        }
+    }
+
+    fn build_attributes(
+        &self,
+        append_from: Option<metrics::MetricAttributes>,
+        attribs: metrics::NewAttributes,
+    ) -> metrics::MetricAttributes {
+        unsafe {
```

Comment:
Unsafe blocks can return values like other blocks - since we already use this pattern of blocks inside of not-unsafe-fns, might as well constrain the scope of the block to just the unsafe part. The meat of the function is still all safe.